kv/util/util.go

```diff
@@ -30,144 +31,226 @@ import (
 	"github.com/m3db/m3x/log"
 )
 
-var errNilStore = errors.New("kv store is nil")
-
-// ValidateFn validates an update from KV.
-type ValidateFn func(interface{}) error
-
-type getValueFn func(kv.Value) (interface{}, error)
+const (
+	// watchRetryInterval is the time interval that the WatchAndUpdate functions
+	// wait when between attempts to reestablish watches.
+	watchRetryInterval = 10 * time.Second
+)
 
-type updateFn func(interface{})
+var (
+	errNilStore  = errors.New("kv store is nil")
+	errNilUpdate = errors.New("update is nil")
+)
 
-// WatchAndUpdateBool sets up a watch with validation for a bool property. Any invalid
-// updates are not applied.
+// WatchAndUpdateBool sets up a watch with validation for a bool property. Any
+// empty, malformed, or invalid updates are not applied.
 func WatchAndUpdateBool(
 	store kv.Store,
 	key string, property *bool,
 	lock sync.Locker,
```

Comment:
nit: lock can be moved to Options as well? the code supports cases with no lock already

Reply:
I thought about this originally, but I couldn't think of any case where you wouldn't want to pass a lock. Since we update the lock in a separate goroutine, without a lock you're introducing a race detection. If we wanted to support a lockless case I would prefer adding functions which performs their writes atomically in a separate PR.

Reply:
Sure, sounds good.